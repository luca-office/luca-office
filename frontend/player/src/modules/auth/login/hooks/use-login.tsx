import {useMutation} from "@apollo/client"
import {useEffect, useState} from "react"
import {useDispatch, useSelector} from "react-redux"
import {SurveyEventType} from "shared/graphql/generated/globalTypes"
import {useSurveyParticipationInfoLazy} from "shared/graphql/hooks"
import {ParticipantData, SurveyInvitation} from "shared/models"
import {updateParticipantDataAction} from "shared/redux/actions/data-action"
import {LucaTFunction, useLucaTranslation} from "shared/translations"
import {isDefined, Option, sendBaseSurveyEvent, useErrorHandler} from "shared/utils"
import {
  UserAccountSurveyParticipationMutation,
  UserAccountSurveyParticipationMutationVariables
} from "../../../../graphql/generated/UserAccountSurveyParticipationMutation"
import {userAccountSurveyParticipationMutation} from "../../../../graphql/mutations"
import {navigateToRouteAction} from "../../../../redux/actions/routing-action"
import {AppState} from "../../../../redux/state/app-state"
import {Route} from "../../../../routes"

export type LoginParticipantData = Omit<ParticipantData, "__typename">

export interface UseLoginHook {
  readonly login: (email: string, password: string) => void
  readonly anonymousLogin: (data: LoginParticipantData) => void
  readonly loginLoading: boolean
  readonly showSignUpModal: boolean
  readonly surveyInvitationData: Option<SurveyInvitation>
  readonly surveyInvitationLoading: boolean
  readonly isOpenParticipationSurvey: boolean
  readonly t: LucaTFunction
  readonly toggleShowSignUpModal: () => void
  readonly navigateToResumption: () => void
}

export const useLogin = (): UseLoginHook => {
  const dispatch = useDispatch()
  const [showSignUpModal, setShowSignUpModal] = useState(false)
  const token = useSelector<AppState, Option<string>>(state => state.data.surveyInvitation.token)

  const {
    surveyParticipationInfoLoading,
    surveyParticipationInfo,
    getSurveyParticipationInfo
  } = useSurveyParticipationInfoLazy()
  const handleError = useErrorHandler()
  const {t} = useLucaTranslation()

  const [participate, {loading: participateLoading}] = useMutation<
    UserAccountSurveyParticipationMutation,
    UserAccountSurveyParticipationMutationVariables
  >(userAccountSurveyParticipationMutation, {
    errorPolicy: "all"
  })

  const surveyInvitationData = surveyParticipationInfo.map(({surveyInvitation}) => surveyInvitation)

  const toggleShowSignUpModal = () => setShowSignUpModal(!showSignUpModal)

  const startProject = (surveyData: SurveyInvitation, participantData?: LoginParticipantData) => {
    dispatch(navigateToRouteAction({routeType: Route.StartProject}))

    if (participantData) {
      sendBaseSurveyEvent({
        surveyId: surveyData.survey.id,
        invitationId: surveyData.id,
        eventType: SurveyEventType.StoreParticipantData,
        data: participantData
      })
    }
  }

  const loginHandler = (email: string, password: string) => {
    surveyInvitationData.forEach(({id}) => {
      participate({
        variables: {surveyInvitationId: id, email, password}
      }).then(({data, errors}) => {
        if (errors === undefined) {
          if (
            data &&
            isDefined(data.userAccountSurveyParticipation?.userAccountId) &&
            isDefined(data.userAccountSurveyParticipation?.userAccount)
          ) {
            const {__typename, ...participantData} = data.userAccountSurveyParticipation?.userAccount

            dispatch(
              updateParticipantDataAction({
                ...participantData,
                __typename: "ParticipantData"
              })
            )

            surveyInvitationData.forEach(surveyData => startProject(surveyData, participantData))
          }
        } else {
          errors.forEach(error => handleError(error))
        }
      })
    })
  }

  const anonLoginHandler = (participantData: LoginParticipantData) => {
    if (participantData) {
      dispatch(
        updateParticipantDataAction({
          ...participantData,
          __typename: "ParticipantData"
        })
      )

      surveyInvitationData.forEach(surveyData => {
        startProject(surveyData, participantData)
      })
    }
  }

  const navigateToResumption = () => dispatch(navigateToRouteAction({routeType: Route.ResumeProject}))

  useEffect(() => {
    token.forEach(invitationToken => getSurveyParticipationInfo(invitationToken))
  }, [token])

  return {
    t,
    surveyInvitationData: surveyInvitationData,
    login: loginHandler,
    anonymousLogin: anonLoginHandler,
    loginLoading: participateLoading,
    showSignUpModal,
    toggleShowSignUpModal: toggleShowSignUpModal,
    surveyInvitationLoading: surveyParticipationInfoLoading,
    navigateToResumption,
    isOpenParticipationSurvey: surveyInvitationData.exists(data => data.survey.isOpenParticipationEnabled)
  }
}
