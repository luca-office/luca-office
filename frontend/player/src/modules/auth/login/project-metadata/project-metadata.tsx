import {css} from "@emotion/react"
import * as React from "react"
import {Columns, Heading, Paper} from "shared/components"
import {HeadingLevel} from "shared/enums"
import {FontWeight, spacingSmall} from "shared/styles"
import {LucaTFunction} from "shared/translations"
import {convertSecondsToMinutes, formatDate} from "shared/utils"
import {MetadataColumn} from "../metadata-column/metadata-column"

interface MetaDataConfig {
  readonly projectName: string
  readonly startsAt: string | null
  readonly endsAt: string | null
  readonly maxDurationInSeconds: number
}
interface Props {
  readonly t: LucaTFunction
  readonly metadata: MetaDataConfig
}

export const ProjectMetadata: React.FC<Props> = ({t, metadata}) => (
  <>
    <Heading level={HeadingLevel.h3} customStyles={styles.label} fontWeight={FontWeight.Bold}>
      {t("project")}
    </Heading>

    <Paper>
      <Columns>
        <MetadataColumn heading={t("title")} value={metadata.projectName} />
        {metadata.startsAt !== null && metadata.endsAt && (
          <MetadataColumn
            heading={t("auth__project_metadata_period")}
            value={`${formatDate(new Date(metadata.startsAt))} - ${formatDate(new Date(metadata.endsAt))}`}
          />
        )}

        <MetadataColumn
          heading={t("auth__project_metadata_processing_time")}
          value={`${convertSecondsToMinutes(metadata.maxDurationInSeconds)} ${t("minutes")}`}
        />
      </Columns>
    </Paper>
  </>
)

const styles = {
  label: css({
    marginBottom: spacingSmall
  })
}
