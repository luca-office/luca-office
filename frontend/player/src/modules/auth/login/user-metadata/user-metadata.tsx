import {css} from "@emotion/react"
import * as React from "react"
import {Columns, Heading, Paper} from "shared/components"
import {HeadingLevel} from "shared/enums"
import {ParticipantData} from "shared/models"
import {FontWeight, spacingHuge, spacingSmall} from "shared/styles"
import {LucaTFunction} from "shared/translations"
import {salutationToLanguageKey} from "shared/utils"
import {MetadataColumn} from "../metadata-column/metadata-column"

interface Props {
  readonly t: LucaTFunction
  readonly participantData: ParticipantData
}

export const UserMetadata: React.FC<Props> = ({t, participantData}) => (
  <div css={styles.userMetadata}>
    <Heading level={HeadingLevel.h3} customStyles={styles.label} fontWeight={FontWeight.Bold}>
      {t("auth__user_information_registered")}
    </Heading>

    <Paper>
      <Columns>
        <MetadataColumn heading={t("first_name")} value={participantData.firstName} />
        <MetadataColumn heading={t("last_name")} value={participantData.lastName} />
        <MetadataColumn
          heading={t("auth__salutation")}
          value={t(salutationToLanguageKey(participantData.salutation))}
        />
      </Columns>
    </Paper>
  </div>
)

const styles = {
  label: css({
    marginBottom: spacingSmall
  }),
  userMetadata: css({
    marginTop: spacingHuge
  })
}
