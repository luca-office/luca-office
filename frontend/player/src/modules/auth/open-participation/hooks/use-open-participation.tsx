import {ApolloError, useMutation} from "@apollo/client"
import * as React from "react"
import {useDispatch} from "react-redux"
import {useSurveyLight} from "shared/graphql/hooks/queries/survey/use-survey-light"
import {SurveyLight} from "shared/models"
import {updateInvitationDataAction} from "shared/redux/actions/data-action"
import {isDefined, Option, parseDateString} from "shared/utils"
import {
  StartOpenParticipationMutation,
  StartOpenParticipationMutationVariables
} from "../../../../graphql/generated/StartOpenParticipationMutation"
import {startOpenParticipationMutation} from "../../../../graphql/mutations"
import {navigateToRouteAction} from "../../../../redux/actions/routing-action"
import {Route} from "../../../../routes"

export interface UseOpenParticipationHook {
  readonly startOpenParticipation: (surveyId: string) => void
  readonly isLoading: boolean
  readonly hasBeenCalled: boolean
  readonly errorCode: Option<string>
  readonly isSurveyLoading: boolean
}

export const useOpenParticipation = (surveyId: Option<UUID>): UseOpenParticipationHook => {
  const dispatch = useDispatch()
  const [errorCode, setErrorCode] = React.useState<Option<string>>(Option.none())

  const onSurveyQueryCompleted = (survey: Option<SurveyLight>) => {
    const isRatingFinalized = survey.exists(survey => survey.isRatingFinalized)
    if (isRatingFinalized) {
      survey.forEach(survey =>
        dispatch(navigateToRouteAction({routeType: Route.ReportLogin, parameters: {surveyId: survey.id}}))
      )
    }
  }

  const {surveyLoading} = useSurveyLight(surveyId.getOrElse(""), undefined, surveyId.isEmpty(), onSurveyQueryCompleted)

  const handleCompletion = (data: StartOpenParticipationMutation) => {
    if (data?.startOpenSurveyParticipation.survey) {
      dispatch(
        updateInvitationDataAction({
          token: Option.of(data.startOpenSurveyParticipation.token),
          invitationId: Option.of(data.startOpenSurveyParticipation.id),
          surveyId: Option.of(data.startOpenSurveyParticipation.survey.id),
          executionType: Option.of(data.startOpenSurveyParticipation.survey.executionType),
          manualSurveyStartedAt: Option.of(
            isDefined(data.startOpenSurveyParticipation.survey.manualPeriod?.start)
              ? parseDateString(data.startOpenSurveyParticipation.survey.manualPeriod!.start)
              : undefined
          )
        })
      )
      dispatch(navigateToRouteAction({routeType: Route.Login}))
    }
  }

  const handleError = (error: ApolloError) => {
    setErrorCode(Option.of(`${error.graphQLErrors[0]?.extensions?.type}`))
  }

  const [startOpenParticipation, {loading, called}] = useMutation<
    StartOpenParticipationMutation,
    StartOpenParticipationMutationVariables
  >(startOpenParticipationMutation, {
    onCompleted: handleCompletion,
    onError: handleError
  })

  return {
    startOpenParticipation: (surveyId: string) => startOpenParticipation({variables: {surveyId: surveyId}}),
    isLoading: loading,
    hasBeenCalled: called,
    errorCode,
    isSurveyLoading: surveyLoading
  }
}
