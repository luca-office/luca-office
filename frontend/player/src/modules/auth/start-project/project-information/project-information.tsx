import * as React from "react"
import {Button, Checkbox, Heading, Paper, Text} from "shared/components"
import {HeadingLevel} from "shared/enums"
import {
  FontWeight,
  primaryColor,
  spacingHuge,
  spacingLarge,
  spacingMedium,
  spacingSmall,
  spacingTiny,
  TextSize
} from "shared/styles"
import {useLucaTranslation} from "shared/translations"

interface Props {
  readonly durationInMinutes: number
  readonly isStartLoading: boolean
  readonly isPrivacyPolicyChecked: boolean
  readonly updateIsPrivacyPolicyChecked: (value: boolean) => void
  readonly onStartClicked: () => void
}

export const ProjectInformation: React.FC<Props> = props => {
  const {
    onStartClicked,
    durationInMinutes,
    isStartLoading,
    isPrivacyPolicyChecked,
    updateIsPrivacyPolicyChecked
  } = props
  const {t} = useLucaTranslation()

  return (
    <div css={styles.startProject}>
      <Heading customStyles={styles.heading} level={HeadingLevel.h3} fontWeight={FontWeight.Bold}>
        {t("project__start_confirm")}
      </Heading>

      <Paper>
        <Text size={TextSize.Medium}>{t("project__start_dialog_text_1", {durationInMinutes})}</Text>

        <Text size={TextSize.Medium} customStyles={styles.privacyPolicyHint}>
          <label css={styles.privacyPolicyHintLabel}>
            <Checkbox
              customStyles={styles.privacyPolicyHintCheckbox}
              checked={isPrivacyPolicyChecked}
              onChange={event => updateIsPrivacyPolicyChecked(event.target.checked)}
            />
            <div>
              {t("project__start_privacy_policy_hint_text_1")}
              &nbsp;
              <a css={styles.primaryTextColor} href={linkUrl1} target="_blank">
                {t("project__start_privacy_policy_hint_link_1")}
              </a>
              {t("project__start_privacy_policy_hint_text_2")}
              &nbsp;
              <a css={styles.primaryTextColor} href={linkUrl2} target="_blank">
                {t("project__start_privacy_policy_hint_link_2")}
              </a>
              &nbsp;
              {t("project__start_privacy_policy_hint_text_3")}
              &nbsp;
              <a css={styles.primaryTextColor} href={linkUrl3} target="_blank">
                {t("project__start_privacy_policy_hint_link_3")}
              </a>
              &nbsp;
              {t("project__start_privacy_policy_hint_text_4")}
            </div>
          </label>
        </Text>

        <Button
          customStyles={styles.button}
          isLoading={isStartLoading}
          disabled={!isPrivacyPolicyChecked}
          onClick={onStartClicked}>
          {t(isStartLoading ? "project__start_loading" : "project__start_now")}
        </Button>
      </Paper>
    </div>
  )
}

const linkUrl1 = "https://luca-office.de/datenschutz-und-nutzungsbedingungen/"
const linkUrl2 = "https://luca-office.de/datenschutz-und-nutzungsbedingungen/"
const linkUrl3 = "https://luca-office.de/datenschutz-und-nutzungsbedingungen/"

const styles = {
  startProject: {
    marginTop: spacingHuge
  },
  heading: {
    marginBottom: spacingTiny
  },
  privacyPolicyHint: {
    marginTop: spacingMedium
  },
  privacyPolicyHintLabel: {
    display: "flex",
    gap: spacingSmall,
    alignItems: "flex-start",
    cursor: "pointer"
  },
  privacyPolicyHintCheckbox: {
    flex: "0 0 auto"
  },
  primaryTextColor: {
    color: primaryColor
  },
  button: {
    width: "50%",
    maxWidth: 300,
    margin: "0 auto",
    marginTop: spacingLarge
  }
}
