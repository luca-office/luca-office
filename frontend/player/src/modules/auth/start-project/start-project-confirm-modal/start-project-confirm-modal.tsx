import {css} from "@emotion/react"
import React from "react"
import {Heading, Modal, Text} from "shared/components"
import {HeadingLevel} from "shared/enums"
import {SurveyInvitation} from "shared/models"
import {Flex, FontWeight, spacingMedium, spacingSmall, TextSize} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {convertSecondsToMinutes} from "shared/utils"

interface Props {
  readonly onConfirm: () => void
  readonly onDismiss: () => void
  readonly surveyInvitation: SurveyInvitation
}

export const StartProjectConfirmModal: React.FC<Props> = ({onConfirm, onDismiss, surveyInvitation}) => {
  const {t} = useLucaTranslation()
  return (
    <Modal
      onConfirm={onConfirm}
      onDismiss={onDismiss}
      customStyles={styles.modal}
      title={`${t("project__start_dialog_title", {title: surveyInvitation.survey.project.name})}`}
      confirmButtonKey="project__start_confirm"
      dismissButtonKey="back_button">
      <Text>
        {t("project__start_dialog_text_confirm_modal", {
          durationInMinutes: convertSecondsToMinutes(surveyInvitation.survey.project.maxDurationInSeconds)
        })}
      </Text>
      <div css={styles.tokenRow}>
        <Heading fontWeight={FontWeight.Bold} level={HeadingLevel.h3}>
          {t("project__start_dialog_text_confirm_modal_code")}
        </Heading>
        <Text customStyles={styles.token} size={TextSize.Medium}>
          {surveyInvitation.token}
        </Text>
      </div>

      <Text customStyles={styles.confirmText}>{t("project__start_dialog_text_2")}</Text>
    </Modal>
  )
}

const styles = {
  modal: css({
    width: "40vw"
  }),
  tokenRow: css(Flex.row, {
    alignItems: "baseline",
    marginTop: spacingMedium,
    marginBottom: spacingMedium
  }),
  token: css({
    marginLeft: spacingSmall
  }),
  confirmText: css({
    marginBottom: spacingMedium
  })
}
