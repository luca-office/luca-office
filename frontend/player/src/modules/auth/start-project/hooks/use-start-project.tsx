import {isEqual} from "lodash-es"
import {Dispatch, SetStateAction, useEffect, useRef, useState} from "react"
import {useDispatch, useSelector} from "react-redux"
import {NotificationSeverity} from "shared/enums"
import {ProjectModuleType, SurveyExecutionType} from "shared/graphql/generated/globalTypes"
import {ProjectModulesQuery} from "shared/graphql/generated/ProjectModulesQuery"
import {useLatestStartedProjectModuleLazy, useSurveyParticipationInfoLazy} from "shared/graphql/hooks"
import {AppNotification, LatestStartedProjectModule, ParticipantData} from "shared/models"
import {OfficeModule} from "shared/models/office-module"
import {updateLatestStartedProjectModuleAction, updateOfficeModules} from "shared/redux/actions/data-action"
import {updateActiveModuleAction, updateNotification} from "shared/redux/actions/ui-action"
import {
  find,
  first,
  isDefined,
  isSurveyParticipationCompleted as checkIsSurveyParticipationCompleted,
  Option,
  sort,
  sortByPosition
} from "shared/utils"
import {useProjectModulesLazy} from "../../../../graphql/hooks"
import {navigateToRouteAction} from "../../../../redux/actions/routing-action"
import {AppState} from "../../../../redux/state/app-state"
import {Route} from "../../../../routes"
import {useSurveyProgress} from "../../../common/hooks/use-survey-progress"
import {toOfficeModule} from "../../util/office-module"

export interface UseStartProjectHook {
  readonly initialized: boolean
  readonly isStartProjectConfirmModalVisible: boolean
  readonly participantData: Option<ParticipantData>
  readonly loading: boolean
  readonly setIsStartProjectConfirmModalVisible: Dispatch<SetStateAction<boolean>>
  readonly loadProjectModules: (projectId: UUID) => void
  readonly isSurveyCompleted: boolean
  readonly isSurveyParticipationCompleted: boolean
  readonly isOpenParticipation: boolean
  readonly surveyId: Option<UUID>
  readonly token: Option<string>
  readonly isPrivacyPolicyChecked: boolean
  readonly updateIsPrivacyPolicyChecked: (value: boolean) => void
}

export const useStartProject = (): UseStartProjectHook => {
  const dispatch = useDispatch()

  const [isPrivacyPolicyChecked, updateIsPrivacyPolicyChecked] = useState(false)
  const tokenRef = useRef<string | null>(null)

  const participantData = useSelector<AppState, Option<ParticipantData>>(state => state.data.common.participantData)
  const executionType = useSelector<AppState, Option<SurveyExecutionType>>(
    state => state.data.surveyInvitation.executionType
  )

  const surveyId = useSelector<AppState, Option<UUID>>(state => state.data.surveyInvitation.surveyId)
  const token = useSelector<AppState, Option<string>>(state => state.data.surveyInvitation.token)

  if (!isEqual(tokenRef.current, token.orNull())) {
    tokenRef.current = token.orNull()
  }

  const {startSpecificModuleById} = useSurveyProgress()

  const {
    surveyParticipationInfoLoading,
    surveyParticipationInfo,
    getSurveyParticipationInfo
  } = useSurveyParticipationInfoLazy()

  const {getLatestStartedProjectModule} = useLatestStartedProjectModuleLazy((latestStartedModuleId, startDate) => {
    const sortedProjectModules = sortByPosition(projectModules.getOrElse([]))
    if (latestStartedModuleId === null) {
      // survey has started but no project module of an attendee -> navigate to first projectModule
      const moduleToBeStarted = first(sortedProjectModules)
      navigateToFirstProjectModule(moduleToBeStarted.map(toOfficeModule))
    } else if (startDate !== null) {
      // navigate to currently active module
      const moduleToBeStarted = find(module => module.id === latestStartedModuleId, sortedProjectModules)
      dispatch(
        updateLatestStartedProjectModuleAction(
          Option.of<LatestStartedProjectModule>({
            projectModuleId: latestStartedModuleId,
            startedAt: startDate?.toISOString()
          })
        )
      )

      navigateToNextProjectModule(moduleToBeStarted.map(toOfficeModule))
    }
  })

  const [isStartProjectConfirmModalVisible, setIsStartProjectConfirmModalVisible] = useState(false)

  const navigateToFirstProjectModule = (firstProjectModule: Option<OfficeModule>) => {
    dispatch(updateActiveModuleAction(firstProjectModule))

    firstProjectModule.forEach(activeModule => {
      dispatch(navigateToRouteAction({routeType: Route.Welcome, parameters: {firstProjectModule: activeModule}}))
    })
  }

  const navigateToNextProjectModule = (nextProjectModule: Option<OfficeModule>) => {
    nextProjectModule.forEach(nextOfficeModule => {
      if (nextOfficeModule.moduleType === ProjectModuleType.Scenario && nextOfficeModule.scenarioId !== null) {
        startSpecificModuleById(ProjectModuleType.Scenario, nextOfficeModule.scenarioId)
      } else if (nextOfficeModule.questionnaireId !== null) {
        startSpecificModuleById(ProjectModuleType.Questionnaire, nextOfficeModule.questionnaireId)
      }
    })
  }

  const onProjectsModulesQueryCompleted = (data: ProjectModulesQuery) => {
    if (data?.projectModules.length > 0) {
      const sortedModules: Array<OfficeModule> = sort(
        projectModule => projectModule.position,
        data.projectModules.map(({moduleType, position, scenarioId, questionnaireId, projectId, id}) => ({
          moduleType,
          position,
          scenarioId,
          questionnaireId,
          projectId,
          id
        }))
      )
      const activeModuleOption = first(sortedModules)

      dispatch(updateOfficeModules(sortedModules))

      if (executionType.exists(type => type === SurveyExecutionType.AutomaticAsynchronous)) {
        navigateToFirstProjectModule(activeModuleOption)
      } else {
        token.forEach(token => {
          getSurveyParticipationInfo(token).then(data => {
            const hasManualSurveyStarted = data.exists(({surveyInvitation}) =>
              isDefined(surveyInvitation.survey.manualPeriod?.start)
            )
            if (hasManualSurveyStarted) {
              surveyId.forEach(surveyId => getLatestStartedProjectModule({variables: {surveyId}}))
            } else {
              dispatch(navigateToRouteAction({routeType: Route.WaitForManualSurveyStart}))
            }
          })
        })
      }
    } else {
      dispatch(
        updateNotification(
          Option.of<AppNotification>({
            severity: NotificationSeverity.Warning,
            messageKey: "scenario__no_scenario_found"
          })
        )
      )
    }
  }

  const onProjectsModulesQueryFailed = () =>
    dispatch(
      updateNotification(
        Option.of<AppNotification>({
          severity: NotificationSeverity.Error,
          messageKey: "scenario__no_scenario_error"
        })
      )
    )

  const {getProjectModules, projectModules, projectModulesLoading, projectModulesCalled} = useProjectModulesLazy({
    onCompleted: onProjectsModulesQueryCompleted,
    onError: onProjectsModulesQueryFailed
  })

  const isSurveyCompleted = surveyParticipationInfo.exists(({surveyInvitation}) => surveyInvitation.survey.isCompleted)
  const isOpenParticipation = surveyParticipationInfo.exists(
    ({surveyInvitation}) => surveyInvitation.isOpenParticipation
  )

  const isSurveyParticipationCompleted = surveyParticipationInfo.exists(({surveyParticipationStatus}) =>
    checkIsSurveyParticipationCompleted(surveyParticipationStatus)
  )

  useEffect(() => {
    if (tokenRef.current !== null) {
      getSurveyParticipationInfo(tokenRef.current)
    }
  }, [tokenRef.current])

  return {
    loadProjectModules: getProjectModules,
    loading: projectModulesLoading || surveyParticipationInfoLoading,
    initialized: projectModulesCalled && !projectModulesLoading,
    isStartProjectConfirmModalVisible,
    setIsStartProjectConfirmModalVisible,
    participantData,
    isSurveyCompleted,
    isSurveyParticipationCompleted,
    isOpenParticipation,
    surveyId,
    token,
    isPrivacyPolicyChecked,
    updateIsPrivacyPolicyChecked
  }
}
