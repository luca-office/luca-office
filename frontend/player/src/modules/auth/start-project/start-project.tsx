import {css} from "@emotion/react"
import * as React from "react"
import {SurveyEventType} from "shared/graphql/generated/globalTypes"
import {SurveyInvitation} from "shared/models"
import {spacingHuge} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {convertSecondsToMinutes, Option, sendBaseSurveyEvent} from "shared/utils"
import {ReportingLoginCard} from "../../reporting"
import {ProjectMetadata} from "../login/project-metadata/project-metadata"
import {UserMetadata} from "../login/user-metadata/user-metadata"
import {useStartProject} from "./hooks/use-start-project"
import {ProjectInformation} from "./project-information/project-information"
import {StartProjectConfirmModal} from "./start-project-confirm-modal/start-project-confirm-modal"

export interface StartProjectProps {
  readonly surveyInvitationData: Option<SurveyInvitation>
}

export const StartProject: React.FC<StartProjectProps> = ({surveyInvitationData}) => {
  const {t} = useLucaTranslation()
  const {
    loadProjectModules,
    loading,
    participantData: participantDataOption,
    isStartProjectConfirmModalVisible,
    setIsStartProjectConfirmModalVisible,
    isSurveyCompleted,
    isSurveyParticipationCompleted,
    isOpenParticipation,
    surveyId: surveyIdOption,
    token: tokenOption,
    isPrivacyPolicyChecked,
    updateIsPrivacyPolicyChecked
  } = useStartProject()

  const startProject = (invitationData: SurveyInvitation) => {
    setIsStartProjectConfirmModalVisible(false)
    sendBaseSurveyEvent({
      surveyId: invitationData.survey.id,
      invitationId: invitationData.id,
      eventType: SurveyEventType.StartProject
    })
    loadProjectModules(invitationData.survey.project.id)
  }

  return surveyInvitationData
    .map(invitationData => {
      return isSurveyCompleted || isSurveyParticipationCompleted
        ? surveyIdOption
            .flatMap(surveyId =>
              tokenOption.map(token => (
                <React.Fragment>
                  <ProjectMetadata
                    t={t}
                    metadata={{
                      endsAt: invitationData.survey.endsAt,
                      startsAt: invitationData.survey.startsAt,
                      maxDurationInSeconds: invitationData.survey.project.maxDurationInSeconds,
                      projectName: invitationData.survey.project.name
                    }}
                  />
                  <ReportingLoginCard
                    customStyles={styles.reportingCard}
                    surveyId={surveyId}
                    token={!isOpenParticipation ? token : undefined}
                  />
                </React.Fragment>
              ))
            )
            .orNull()
        : participantDataOption
            .map(participantData => (
              <React.Fragment>
                <ProjectMetadata
                  t={t}
                  metadata={{
                    endsAt: invitationData.survey.endsAt,
                    startsAt: invitationData.survey.startsAt,
                    maxDurationInSeconds: invitationData.survey.project.maxDurationInSeconds,
                    projectName: invitationData.survey.project.name
                  }}
                />
                <UserMetadata t={t} participantData={participantData} />
                <ProjectInformation
                  durationInMinutes={convertSecondsToMinutes(invitationData.survey.project.maxDurationInSeconds)}
                  isStartLoading={loading}
                  isPrivacyPolicyChecked={isPrivacyPolicyChecked}
                  updateIsPrivacyPolicyChecked={updateIsPrivacyPolicyChecked}
                  onStartClicked={() => setIsStartProjectConfirmModalVisible(true)}
                />
                {isStartProjectConfirmModalVisible && (
                  <StartProjectConfirmModal
                    surveyInvitation={invitationData}
                    onConfirm={() => startProject(invitationData)}
                    onDismiss={() => setIsStartProjectConfirmModalVisible(false)}
                  />
                )}
              </React.Fragment>
            ))
            .orNull()
    })
    .orNull()
}

const styles = {
  reportingCard: css({
    marginTop: spacingHuge
  })
}
