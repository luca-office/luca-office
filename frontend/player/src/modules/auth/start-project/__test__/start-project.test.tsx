import {mount} from "enzyme"
import * as React from "react"
import {Provider} from "react-redux"
import {create} from "react-test-renderer"
import {surveyIdMock, surveyInvitationsMock} from "shared/graphql/__mocks__"
import {ParticipantData} from "shared/models"
import {Option} from "shared/utils"
import {fakeStore} from "sharedTests/redux/fake-store"
import {surveyInvitationMock} from "../../../../graphql/__mocks__"
import {Salutation} from "../../../../graphql/generated/globalTypes"
import {initialAppState} from "../../../../redux/state/app-state"
import {ProjectMetadata} from "../../login/project-metadata/project-metadata"
import {UserMetadata} from "../../login/user-metadata/user-metadata"
import * as startProjectHook from "../hooks/use-start-project"
import {UseStartProjectHook} from "../hooks/use-start-project"
import {ProjectInformation} from "../project-information/project-information"
import {StartProject, StartProjectProps} from "../start-project"
import {StartProjectConfirmModal} from "../start-project-confirm-modal/start-project-confirm-modal"

const defaultProps: StartProjectProps = {
  surveyInvitationData: Option.none()
}

const getComponent = (props?: Partial<React.PropsWithChildren<StartProjectProps>>) => (
  <Provider store={fakeStore(initialAppState(null))}>
    <StartProject {...{...defaultProps, ...props}} />
  </Provider>
)

const participantDataMock: ParticipantData = {
  __typename: "ParticipantData",
  firstName: "Max",
  lastName: "Mustermann",
  salutation: Salutation.Mr
}

const hookValuesDefault: UseStartProjectHook = {
  initialized: false,
  isStartProjectConfirmModalVisible: false,
  participantData: Option.of(participantDataMock),
  loading: false,
  setIsStartProjectConfirmModalVisible: jest.fn(),
  loadProjectModules: jest.fn(),
  isSurveyCompleted: false,
  isSurveyParticipationCompleted: false,
  isOpenParticipation: false,
  surveyId: Option.of(surveyIdMock),
  token: Option.of(surveyInvitationsMock[0].token),
  isPrivacyPolicyChecked: false,
  updateIsPrivacyPolicyChecked: jest.fn()
}

const stateSpy = jest.spyOn(startProjectHook, "useStartProject")

describe("start-project", () => {
  it("renders correctly", () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure with logged in user and not survey invitation data", () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = getComponent()
    const tree = mount(component)

    expect(tree.find(ProjectMetadata)).toHaveLength(0)
    expect(tree.find(UserMetadata)).toHaveLength(0)
    expect(tree.find(ProjectInformation)).toHaveLength(0)
  })

  it("has correct structure with logged in user and survey invitation data", () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = getComponent({surveyInvitationData: Option.of(surveyInvitationMock)})
    const tree = mount(component)

    expect(tree.find(ProjectMetadata)).toHaveLength(1)
    expect(tree.find(UserMetadata)).toHaveLength(1)
    expect(tree.find(ProjectInformation)).toHaveLength(1)
    expect(tree.find(ProjectInformation).prop("isStartLoading")).toBe(false)
  })
  it("has correct structure with logged in user and start is loading", () => {
    stateSpy.mockReturnValue({...hookValuesDefault, loading: true})
    const component = getComponent({surveyInvitationData: Option.of(surveyInvitationMock)})
    const tree = mount(component)

    expect(tree.find(ProjectMetadata)).toHaveLength(1)
    expect(tree.find(UserMetadata)).toHaveLength(1)
    expect(tree.find(ProjectInformation)).toHaveLength(1)
    expect(tree.find(ProjectInformation).prop("isStartLoading")).toBe(true)
  })
  it("has correct structure with confirm modal", () => {
    stateSpy.mockReturnValue({...hookValuesDefault, isStartProjectConfirmModalVisible: true})
    const component = getComponent({surveyInvitationData: Option.of(surveyInvitationMock)})
    const tree = mount(component)

    expect(tree.find(ProjectMetadata)).toHaveLength(1)
    expect(tree.find(UserMetadata)).toHaveLength(1)
    expect(tree.find(ProjectInformation)).toHaveLength(1)
    expect(tree.find(StartProjectConfirmModal)).toHaveLength(1)
  })
})
