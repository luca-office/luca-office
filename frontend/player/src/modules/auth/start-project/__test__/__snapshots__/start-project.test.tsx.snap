// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start-project renders correctly 1`] = `null`;
